crates/contracts/solidity/Trader.sol

```diff
@@ -13,78 +13,72 @@ contract Trader {
-    function settle(
-        address[] calldata tokens,
-        uint256[] calldata clearingPrices,
-        Interaction[][3] calldata interactions,
-        uint256 mint
-    ) external returns (
-        uint256 gasUsed,
-        int256[] memory traderBalances,
-        int256[] memory settlementBalances
-    ) {
-        if (mint != 0) {
-            IPhonyERC20(tokens[0]).mintPhonyTokens(address(this), mint);
-        }
-        // Make sure to reset the approval before setting a new one - some
-        // popular tokens (like Tether USD) require this.
-        IERC20(tokens[0]).safeApprove(address(SETTLEMENT.vaultRelayer()), 0);
-        IERC20(tokens[0]).safeApprove(address(SETTLEMENT.vaultRelayer()), type(uint256).max);
+    /// @dev Address where the original code for the trader implementation is
+    /// expected to be. Use 0x10000 as its the first "valid" address, since
+    /// addresses up to 0xffff are reserved for pre-compiles.
+    /// This is used to proxy calls to the original implementation in case
+    /// the trader is actually a smart contract.
+    address constant private TRADER_IMPL = address(0x10000);
 
-        traderBalances = new int256[](tokens.length);
-        settlementBalances = new int256[](tokens.length);
-        for (uint256 i; i < tokens.length; ++i) {
-            traderBalances[i] = -IERC20(tokens[i]).balanceOf(address(this)).toInt();
-            settlementBalances[i] = -IERC20(tokens[i]).balanceOf(address(SETTLEMENT)).toInt();
-        }
+    /// @dev Flag that ensures that `prepareSwap` gets called exactly once to
+    /// prevent custom interaction from calling it.
+    bool private _alreadyCalled = false;
 
-        Trade[] memory trades = new Trade[](1);
-        trades[0] = Trade({
-            sellTokenIndex: 0,
-            buyTokenIndex: tokens.length - 1,
-            receiver: address(0),
-            sellAmount: clearingPrices[tokens.length - 1],
-            buyAmount: clearingPrices[0],
-            validTo: type(uint32).max,
-            appData: bytes32(0),
-            feeAmount: 0,
-            flags: 0x40, // EIP-1271
-            executedAmount: 0,
-            signature: abi.encodePacked(address(this))
-        });
+    // The `Trader` contract gets deployed on the `from` address of the quote.
+    // Since the `from` address might be a safe or other smart contract we still
+    // need to make the `Trader` behave as the original `from` would have in
+    // case some custom interactions rely on that behvavior.
+    // To do that we simply implement fallback handlers that do delegate calls
+    // to the original implementation.
+    fallback() external payable {
+        _fallback();
+    }
+    receive() external payable {
+        _fallback();
+    }
+    function _fallback() private {
+        bytes memory rdata = TRADER_IMPL.doDelegatecall(msg.data);
+        assembly { return(add(rdata, 32), mload(rdata)) }
+    }
```

Comment:
This might push receiving ETH over the 2300 gas stipend (which would cause buying ETH to fail in simulation). I would actually on `receive()` just do nothing (in fact, SC wallets that don't "do nothing" aren't properly supported by the Settlement contract anyway).

Concretely, I would change this to:
```suggestion
    fallback() external payable {
        bytes memory rdata = TRADER_IMPL.doDelegatecall(msg.data);
        assembly { return(add(rdata, 32), mload(rdata)) }
    }
    receive() external payable {}
```